adapters/insticator/insticator.go

```diff
@@ -0,0 +1,335 @@
+		}
+	}
+
+	// check if the adUnitId is not empty
+	if insticatorExt.AdUnitId == "" {
+		return openrtb2.Imp{}, "", &errortypes.BadInput{
+			Message: "Missing adUnitId",
+		}
+	}
+
+	// check if the publisherId is not empty
+	if insticatorExt.PublisherId == "" {
+		return openrtb2.Imp{}, insticatorExt.AdUnitId, &errortypes.BadInput{
+			Message: "Missing publisherId",
+		}
+	}
+
+	// Directly construct the impExt
+	impExt := ext{
+		Insticator: impInsticatorExt{
+			AdUnitId:    insticatorExt.AdUnitId,
+			PublisherId: insticatorExt.PublisherId,
+		},
+	}
+
+	impExtJSON, err := json.Marshal(impExt)
+	if err != nil {
+		return openrtb2.Imp{}, "", &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	imp.Ext = impExtJSON
+
+	// Validate Video if it exists
+	if imp.Video != nil {
+		if err := validateVideoParams(imp.Video); err != nil {
+			return openrtb2.Imp{}, insticatorExt.AdUnitId, &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
+	}
+
+	// Return the imp, AdUnitId, and no error
+	return imp, insticatorExt.AdUnitId, nil
+}
+
+func makeReqExt(request *openrtb2.BidRequest) ([]byte, error) {
+	var reqExt reqExt
+
+	if len(request.Ext) > 0 {
+		if err := json.Unmarshal(request.Ext, &reqExt); err != nil {
+			return nil, err
+		}
+	}
+
+	if reqExt.Insticator == nil {
+		reqExt.Insticator = &reqInsticatorExt{}
+	}
+
+	if reqExt.Insticator.Caller == nil {
+		reqExt.Insticator.Caller = make([]insticatorCaller, 0)
+	}
+
+	reqExt.Insticator.Caller = append(reqExt.Insticator.Caller, caller)
+
+	return json.Marshal(reqExt)
+}
+
+func resolveBidFloor(bidFloor float64, bidFloorCur string, reqInfo *adapters.ExtraRequestInfo) (float64, error) {
+	if bidFloor > 0 && bidFloorCur != "" && strings.ToUpper(bidFloorCur) != "USD" {
+		floor, err := reqInfo.ConvertCurrency(bidFloor, bidFloorCur, "USD")
+		return roundTo4Decimals(floor), err
+	}
+
+	return bidFloor, nil
+}
+
+// roundTo4Decimals function
+func roundTo4Decimals(amount float64) float64 {
+	return math.Round(amount*10000) / 10000
+}
+
+func validateVideoParams(video *openrtb2.Video) error {
+	videoCopy := *video
+	if (videoCopy.W == nil || *videoCopy.W == 0) ||
+		(videoCopy.H == nil || *videoCopy.H == 0) ||
+		videoCopy.MIMEs == nil {
+
+		return &errortypes.BadInput{
+			Message: "One or more invalid or missing video field(s) w, h, mimes",
+		}
+	}
+
+	return nil
+}
+
+// populatePublisherId function populates site.publisher.id or app.publisher.id
+func populatePublisherId(imp *openrtb2.Imp, request *openrtb2.BidRequest) error {
+	var ext ext
+
+	// Unmarshal the imp extension to get the publisher ID
+	if err := json.Unmarshal(imp.Ext, &ext); err != nil {
+		return &errortypes.BadInput{Message: "Error unmarshalling imp extension"}
+	}
+
+	// Populate site.publisher.id if request.Site is not nil
+	if request.Site != nil {
+		if request.Site.Publisher == nil {
+			request.Site.Publisher = &openrtb2.Publisher{}
+		}
+		request.Site.Publisher.ID = ext.Insticator.PublisherId
```

Comment:
This is shared memory. You need to make a shallow copy of Site and Site.Publisher first if they already exist.